### ISSUE TITLE
Recent update of GNU grep breaks Perl regular expressions

### ISSUE THREAD

**Issue by vwkd** [Homebrew/homebrew-core#128277]
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```
HOMEBREW_VERSION: 4.0.13
ORIGIN: https://github.com/Homebrew/brew
HEAD: 9b60fcf7ac3f82fa589a64091376d0663262f8c6
Last commit: 22 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 25be650e536c44d8294ad15be81243c14aeff69d
Core tap last commit: 8 weeks ago
Core tap branch: master
Core tap JSON: 13 Apr 16:21 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 2.6.10 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.10_1/bin/ruby
CPU: quad-core 64-bit skylake
Clang: 14.0.0 build 1400
Git: 2.40.0 => /usr/local/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 12.6.4-x86_64
CLT: 14.2.0.0.1.1668646533
Xcode: 14.2
```

```
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  six

Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting your PATH for example like so:
  echo 'export PATH="/usr/local/sbin:$PATH"' >> ~/.zshrc

Warning: The directory Xcode is reportedly installed to doesn't exist:
  /Applications/Xcode.app/Contents/Developer
You may need to `xcode-select` the proper path if you have moved Xcode.
```

### Verification

- [ ] My "`brew doctor` output" says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.
- [X] I searched for recent similar issues at https://github.com/Homebrew/homebrew-core/issues?q=is%3Aissue and found no duplicates.

### What were you trying to do (and why)?

I recently upgraded GNU grep to the latest version `3.10` from some previous version (I don't remember which version it was before. Does anyone know how to find it?) and running some Pearl regular expression. 

```
echo "foo bar" | ggrep -P 'foo(?![ \d])'
```

### What happened (include all command output)?

The output prints `foo bar`.

### What did you expect to happen?

No output.

Verify also on [regex101.com](https://regex101.com/r/kTQRft/1).

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
1. `brew install grep`
2. `echo "foo bar" | ggrep -P 'foo(?![ \d])'`
```

**Comment by SMillerDev** [comment]
Please report this to the developers of Grep since this seems unrelated to brew.

**Comment by ZhongRuoyu** [comment]
Indeed this is a bug on `grep`'s side. Fortunately, it has been fixed upstream few days ago, after the 3.1.0 release. (See [`NEWS`](https://git.savannah.gnu.org/gitweb/?p=grep.git;a=blob;f=NEWS;h=3fdf4fe33df451612d2e1a8fd0b38ac6cb0a9b4a;hb=refs/heads/master#l5).) The fix should be in `grep`'s next release, but if you want it now, I'm adding a `head` spec to `grep` in #128292 so that you may try `brew install --head grep` while waiting for the next stable release.

**Comment by vwkd** [comment]
Thanks! Unfortunately, `brew install --head grep` fails with an error relating to `autoconf`.

```
==> Pouring autoconf--2.71.monterey.bottle.tar.gz
Error: No such file or directory @ rb_sysopen - /Users/[username]/Library/Caches/Homebrew/downloads/1a46ace543df66c6a0ac223056cbd805a93c209c650c12b791eb16a0d7354359--autoconf--2.71.monterey.bottle.tar.gz
```

It seems, the issue was introduced by grep 3.8 switching to PCRE2. The latest homebrew commit of grep 3.7 seems to be [`7e79b142767efb12b1d17f4f1b4242c712984b5a`](https://github.com/Homebrew/homebrew-core/blob/7e79b142767efb12b1d17f4f1b4242c712984b5a/Formula/grep.rb).

I was able to downgrade to grep 3.7 using the following steps:
1. `brew unpin grep`
2. `cd "$(brew --repo homebrew/core)"`
3. `git switch -c grep-3.7 7e79b142767efb12b1d17f4f1b4242c712984b5a`
4. `brew install -s Formula/grep.rb`
5. `brew pin grep`
6. `git switch master`

**Comment by ZhongRuoyu** [comment]
This seems to be an issue with API install. You should be able to install it like this, without having to downgrade `grep`:

```
export HOMEBREW_NO_INSTALL_FROM_API=1
brew update
brew install --head grep
```

**Comment by vwkd** [comment]
Thanks, this solved the `autoconf` issue. Unfortunately, building `grep` from source fails for me with the error below.

I'm in no hurry to upgrade, so downgrading seems like a simpler solution for now.

```
==> Installing grep --HEAD
==> ./bootstrap
==> ./configure --disable-nls --prefix=/usr/local/Cellar/grep/HEAD-6e7253d --infod
==> make
Last 15 lines from /Users/[username]/Library/Logs/Homebrew/grep/03.make:
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: libgreputils.a(libgreputils_a-threadlib.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: libgreputils.a(libgreputils_a-unistd.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: libgreputils.a(libgreputils_a-u8-uctomb.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: libgreputils.a(libgreputils_a-wctype-h.o) has no symbols
Making all in doc
  GEN      grep.1
Updating ./version.texi
  MAKEINFO grep.info
grep.texi:17: warning: unrecognized encoding name `UTF-8'.
grep.texi:2229: Unknown command `frenchspacing'.
grep.texi:2273: Unknown command `frenchspacing'.
makeinfo: Removing output file `grep.info' due to errors; use --force to preserve.
make[2]: *** [grep.info] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2

READ THIS: https://docs.brew.sh/Troubleshooting

grep's formula was built from an unstable upstream --HEAD.
This build failure is expected behaviour.
Do not create issues about this on Homebrew's GitHub repositories.
Any opened issues will be immediately closed without response.
Do not ask for help from Homebrew or its maintainers on social media.
You may ask for help in Homebrew's discussions but are unlikely to receive a response.
Try to figure out the problem yourself and submit a fix as a pull request.
We will review it but may or may not accept it.
```

**Comment by ZhongRuoyu** [comment]
> grep.texi:2229: Unknown command `frenchspacing'.

That seems to be an issue with the `texinfo` shipped by macOS (prior to Ventura), and should be fixed by #128815.

**Comment by ZhongRuoyu** [comment]
Sorry for the back and forth. If you do this again you should be able to install the latest `grep`:

```
export HOMEBREW_NO_INSTALL_FROM_API=1
brew update
brew install --head grep
```
